### PR TITLE
Pre-register mod metadata during Windows / .deb installation.

### LINKED ISSUE #12764: Automatically register mod metadata during installation.
It is fairly straightforward to make the Windows and Linux installers pre-register the default mod metadata during install (so that they are available for fast-switching without having to be manually launched the first time), and my [utility-registration](https://github.com/pchote/OpenRA/tree/utility-registration) branch adds this.

This branch isn't mergable in its current state, because writing user-specific files for only the current user from an installer is very bad practice.

Once #10374 has been solved we can switch the metadata files over to the shared data dir and this proof of concept should be mergable as-is.

### PR DESCRIPTION
**Pull request by pchote**
Requiring users to switch to a mod manually before it is available for switching caused a lot of confusion when `release-20170421` and `release-20170527` were first released.  This PR fixes the issue for Windows and Linux (just deb here, but downstream packaging is also expected to adopt this) by pre-registering mods during installation.

A new "system support" directory is defined at:
* Windows: `C:\ProgramData\OpenRA\` ([spec](https://msdn.microsoft.com/en-us/library/s2esdf4x(v=vs.100).aspx))
* Linux: `/var/games/openra/` ([spec](http://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s07.html))
* macOS: `/Library/Application Support/OpenRA/`

and a `ModMetadata` subdirectory is created and populated for ra/td/d2k.  These are left read-only for normal users to avoid a malicious user from changing the `LaunchPath` to something nasty.

Closes #12764 and implements the first steps towards #10374.

### CODE REVIEW

File: OpenRA.Game/ExternalMods.cs [hunk 1]
```diff
@@ -50,21 +52,24 @@ public ExternalMods(string launchPath)
...
+					return;
+
+				foreach (var path in Directory.GetFiles(supportPath, "*.yaml"))
 				{
-					Log.Write("debug", "Failed to parse mod metadata file '{0}'", path);
-					Log.Write("debug", e.ToString());
+					try
+					{
+						var yaml = MiniYaml.FromStream(File.OpenRead(path), path).First().Value;
+						LoadMod(yaml);
```
Comment (rob-v) [inline OpenRA.Game/ExternalMods.cs]: In case "Support" directory exists, this will be executed twice for the same path (SystemSupportDir = SupportDir).

File: OpenRA.Game/ExternalMods.cs [hunk 2]
```diff
@@ -107,53 +113,96 @@ internal void Register(Manifest mod)
...
-				Directory.CreateDirectory(supportPath);
-				File.WriteAllLines(Path.Combine(supportPath, key + ".yaml"), yaml.ToLines(false).ToArray());
-			}
-			catch (Exception e)
-			{
-				Log.Write("debug", "Failed to register currrent mod metadata");
-				Log.Write("debug", e.ToString());
-			}
+			// Make sure the mod is available for this session, even if saving it fails
+			LoadMod(yaml.First().Value);
```
Comment (rob-v) [inline OpenRA.Game/ExternalMods.cs]: This was moved out of try block, can it fail and crash now?

File: OpenRA.Game/ExternalMods.cs [hunk 3]
```diff
@@ -107,53 +113,96 @@ internal void Register(Manifest mod)
...
+				locations.Add(Platform.SystemSupportDir);
+
+			if (registration.HasFlag(ModRegistration.User))
+				locations.Add(Platform.SupportDir);
 
-			if (toRemove != null)
-				foreach (var r in toRemove)
-					mods.Remove(r);
+			var key = ExternalMod.MakeKey(mod);
+			mods.Remove(key);
```
Comment (rob-v) [inline OpenRA.Game/ExternalMods.cs]: I propose to add check if mod exists to avoid crash.

File: OpenRA.Game/Platform.cs [hunk 4]
```diff
@@ -95,6 +99,33 @@ static string GetSupportDir()
...
+		{
+			// Use a local directory in the game root if it exists (shared with the system support dir)
+			var localSupportDir = Path.Combine(GameDir, "Support");
+			if (Directory.Exists(localSupportDir))
+				return localSupportDir + Path.DirectorySeparatorChar;
+
+			switch (CurrentPlatform)
+			{
+				case PlatformType.Windows:
+					return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), "OpenRA");
```
Comment (rob-v) [inline OpenRA.Game/Platform.cs]: \+ Path.DirectorySeparatorChar

File: OpenRA.Game/Platform.cs [hunk 5]
```diff
@@ -95,6 +99,33 @@ static string GetSupportDir()
...
+				case PlatformType.OSX:
+					return "/Library/Application Support/OpenRA/";
+				default:
+					return "/var/games/openra";
```
Comment (rob-v) [inline OpenRA.Game/Platform.cs]: / not needed?

**Review by rob-v — COMMENTED** [review]
Not tested yet ingame, small issues. Should I test the installation on Win? I don't know how to create new setup.

File: OpenRA.Game/ExternalMods.cs (hunk 2 above)
Comment (pchote) [inline OpenRA.Game/ExternalMods.cs]: This is parsing yaml that we have defined just above, and should never be invalid.  It shouldn't be possible to crash here on anything that didn't crash somewhere else earlier.

File: OpenRA.Game/ExternalMods.cs (hunk 3 above)
Comment (pchote) [inline OpenRA.Game/ExternalMods.cs]: This simply returns false if the mod doesn't exist.  The only way it can crash is if `key` is null, which isn't possible.

**Comment by pchote** [comment]
Updated.

I have also built test installers (which I forgot to share first time around):
* https://www.dropbox.com/s/xetqxm17e2m57kn/openra_test.20170626_all.deb?dl=0
* https://www.dropbox.com/s/qluwgvbw859i5wm/OpenRA-test-20170626.exe?dl=0

File: OpenRA.Game/ExternalMods.cs [hunk 6]
```diff
@@ -107,53 +120,96 @@ internal void Register(Manifest mod)
...
+
+				if (toRemove != null)
+					foreach (var r in toRemove)
+						mods.Remove(r);
 			}
+		}
+
+		internal void Unregister(Manifest mod, ModRegistration registration)
+		{
+			var locations = new List<string>();
```
Comment (rob-v) [inline OpenRA.Game/ExternalMods.cs]: nit: you called it sources in ExternalMods

File: OpenRA.Game/UtilityCommands/UnregisterModCommand.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+		void IUtilityCommand.Run(Utility utility, string[] args)
+		{
+			ModRegistration type = 0;
+			if (args[1] == "system" || args[1] == "both")
+				type |= ModRegistration.System;
+
+			if (args[1] == "user" || args[1] == "both")
+				type |= ModRegistration.User;
+
+			new ExternalMods(args[1]).Unregister(utility.ModData.Manifest, type);
```
Comment (rob-v) [inline OpenRA.Game/UtilityCommands/UnregisterModCommand.cs]: args[1] isn't launchPath here

File: OpenRA.Game/ExternalMods.cs [hunk 8]
```diff
@@ -107,53 +120,96 @@ internal void Register(Manifest mod)
...
+			if (registration.HasFlag(ModRegistration.User))
+				locations.Add(Platform.SupportDir);
 
-			if (toRemove != null)
-				foreach (var r in toRemove)
-					mods.Remove(r);
+			var key = ExternalMod.MakeKey(mod);
+			mods.Remove(key);
+
+			foreach (var source in locations)
```
Comment (rob-v) [inline OpenRA.Game/ExternalMods.cs]: better locations(sources).Distinct()

File: OpenRA.Game/ExternalMods.cs [hunk 9]
```diff
@@ -107,53 +120,96 @@ internal void Register(Manifest mod)
...
-			}
+			// Make sure the mod is available for this session, even if saving it fails
+			LoadMod(yaml.First().Value);
 
-			// Clean up stale mod registrations:
-			//  - LaunchPath no longer exists (uninstalled)
-			//  - LaunchPath and mod match the current mod, but version differs (newer version installed on top)
-			List<string> toRemove = null;
-			foreach (var kv in mods)
+			foreach (var source in sources)
```
Comment (rob-v) [inline OpenRA.Game/ExternalMods.cs]: better sources.Distinct()

**Review by rob-v — COMMENTED** [review]
small issues

**Comment by pchote** [comment]
Fixed.

File: OpenRA.Game/ExternalMods.cs [hunk 10]
```diff
@@ -80,14 +81,19 @@ void LoadMod(MiniYaml yaml)
 					mod.Icon = sheetBuilder.Add(bitmap);
 			}
 
-			mods.Add(ExternalMod.MakeKey(mod), mod);
+			if (File.Exists(mod.LaunchPath))
+				mods[ExternalMod.MakeKey(mod)] = mod;
```
Comment (rob-v) [inline OpenRA.Game/ExternalMods.cs]: Side effect of this is that deleted mods won't be removed as it loops mods. It could loop .yaml files instead.

File: OpenRA.Game/ExternalMods.cs [hunk 11]
```diff
@@ -107,53 +113,98 @@ internal void Register(Manifest mod)
...
-					try
-					{
-						File.Delete(path);
-					}
-					catch (Exception e)
+				// Clean up stale mod registrations:
+				//  - LaunchPath no longer exists (uninstalled)
+				//  - LaunchPath and mod match the current mod, but version differs (newer version installed on top)
+				List<string> toRemove = null;
+				foreach (var kv in mods)
```
Comment (rob-v) [inline OpenRA.Game/ExternalMods.cs]: Deleted mods are now not included in mods.

File: packaging/windows/OpenRA.nsi [hunk 12]
```diff
@@ -150,6 +150,12 @@ Section "Game" GAME
 	${GetSize} "$INSTDIR" "/S=0K" $0 $1 $2
 	IntFmt $0 "0x%08X" $0
 	WriteRegDWORD HKLM "Software\Microsoft\Windows\CurrentVersion\Uninstall\OpenRA${SUFFIX}" "EstimatedSize" "$0"
+
+	SetShellVarContext all
+	CreateDirectory "$APPDATA\OpenRA\ModMetadata"
+	nsExec::ExecToLog '"$INSTDIR\OpenRA.Utility.exe" ra --register-mod "$INSTDIR\OpenRA.exe" system'
+	nsExec::ExecToLog '"$INSTDIR\OpenRA.Utility.exe" cnc --register-mod "$INSTDIR\OpenRA.exe" system'
+	nsExec::ExecToLog '"$INSTDIR\OpenRA.Utility.exe" d2k --register-mod "$INSTDIR\OpenRA.exe" system'
```
Comment (rob-v) [inline packaging/windows/OpenRA.nsi]: It needs to be $INSTDIR\OpenRA.Game.exe.

File: OpenRA.Game/ExternalMods.cs [hunk 13]
```diff
@@ -38,33 +40,32 @@ public class ExternalMods : IReadOnlyDictionary<string, ExternalMod>
...
 			{
-				try
-				{
-					var yaml = MiniYaml.FromStream(File.OpenRead(path), path).First().Value;
-					LoadMod(yaml);
-				}
-				catch (Exception e)
+				var supportPath = Path.Combine(source, "ModMetadata");
+				if (!Directory.Exists(supportPath))
+					return;
```
Comment (rob-v) [inline OpenRA.Game/ExternalMods.cs]: It needs to be continue instead of return.

**Review by rob-v — COMMENTED** [review]
Small bugs.
Utility can be used to register only standard mods, could it work for all mods? as e.g. cd --register-mod is refused.

File: packaging/windows/OpenRA.nsi (hunk 12 above)
Comment (pchote) [inline packaging/windows/OpenRA.nsi]: No, OpenRA.exe is correct.  Calling OpenRA.Game.exe directly would then remove the crash dialog and mod window icon setting.

**Comment by pchote** [comment]
> e.g. cd --register-mod is refused.

Refused how?  I see no reason why this shouldn't work.

**Comment by rob-v** [comment]
There is somewhere check for valid mods: The available mods are: cnc, d2k, modcontent, ra

File: packaging/windows/OpenRA.nsi (hunk 12 above)
Comment (rob-v) [inline packaging/windows/OpenRA.nsi]: OpenRA.exe doesn't exist, so it won't be registered now as you check for launch path.

**Comment by pchote** [comment]
The `--register-mod` command won't be available in CD until they update their engine version to include this PR after merging.  The "official mods" engine installation will not be able to see any mod installations, and similarly any mod installations will not be able to see the "official mods" installation.

File: packaging/windows/OpenRA.nsi (hunk 12 above)
Comment (pchote) [inline packaging/windows/OpenRA.nsi]: Ah right, we renamed these to mod-specific launchers.

**Comment by pchote** [comment]
> Side effect of this is that deleted mods won't be removed as it loops mods. It could loop .yaml files instead.

I ended up having to rewriting most of the ExternalMods registration and cleanup code to address this in a way that I was happy with.

I have put some effort into testing the various edge cases that may happen during normal game launches, but ran out of time to test the utility commands and packaged versions.  I'll try and do that tomorrow if you don't spot anything first.

**Comment by pchote** [comment]
I strongly suggest reviewing by commit now, using `w=1` on the "Read mod registrations from system support dir." commit.

File: OpenRA.Game/Game.cs [hunk 14]
```diff
@@ -338,8 +338,25 @@ internal static void Initialize(Arguments args)
...
+				// Sanitize input from platform-specific launchers
+				// Process.Start requires paths to not be quoted, even if they contain spaces
+				if (launchPath.First() == '"' && launchPath.Last() == '"')
+					launchPath = launchPath.Substring(1, launchPath.Length - 2);
+
+				ExternalMods.Register(Mods[modID], launchPath, ModRegistration.User);
+
+				ExternalMod activeMod;
+				if (ExternalMods.TryGetValue(ExternalMod.MakeKey(Mods[modID]), out activeMod))
+					ExternalMods.ClearInvalidRegistrations(activeMod, ModRegistration.User);
```
Comment (rob-v) [inline OpenRA.Game/Game.cs]: ClearInvalidRegistrations is called only once atm - ModRegistration.User and an activeMod, isn't call for System missing?

File: OpenRA.Game/ExternalMods.cs [hunk 15]
```diff
@@ -107,41 +111,87 @@ internal void Register(Manifest mod)
...
+					string modKey = null;
+					try
+					{
+						var yaml = MiniYaml.FromStream(File.OpenRead(path), path).First().Value;
+						var m = FieldLoader.Load<ExternalMod>(yaml);
+						modKey = ExternalMod.MakeKey(m);
+
+						// Continue to the next entry if this one is valid
+						if (File.Exists(m.LaunchPath) && Path.GetFileNameWithoutExtension(path) == modKey &&
+							!(activeMod != null && m.LaunchPath == activeMod.LaunchPath && m.Id == activeMod.Id && m.Version != activeMod.Version))
```
Comment (rob-v) [inline OpenRA.Game/ExternalMods.cs]: only note: Can this happen? Register call before ClearInvalidRegistrations overwrites LaunchPath, Id and Version.

**Review by rob-v — COMMENTED** [review]
Some questions.

File: OpenRA.Game/ExternalMods.cs (hunk 15 above)
Comment (pchote) [inline OpenRA.Game/ExternalMods.cs]: Consider a person on windows installing `release-20170421`.  This will write a metadata file ra-release-20170421.yaml pointing to `('ra', 'release-20170421', 'C:\Program Files\OpenRA\OpenRA.exe')`.  They then install `release-20170527` over the top without uninstalling the old version first, so we get a new metadata file ra-release-20170527.yaml pointing to `('ra', 'release-20170527', 'C:\Program Files\OpenRA\OpenRA.exe')` written alongside it.

When they then run the game the `release-20175027` metadata will update itself, but the `release-20170421` entry will not be touched until this line runs and discovers that we have a different version of RA claiming to run at the same launch path as the current version of RA, which we know is impossible.

We can't ignore all except the current version of RA, because we need to support switching between ra release and ra playtest installed at `C:\Program Files\OpenRA (playtest)\RedAlert.exe`.

File: OpenRA.Game/Game.cs (hunk 14 above)
Comment (pchote) [inline OpenRA.Game/Game.cs]: This is intentional.  The system metadata is managed by the package manager / installer, and it would be bogus to delete arbitrary files from there ourselves.

Comment (pchote) [inline OpenRA.Game/Game.cs]: I guess i'll have to add another utility command to clear invalid System registrations during the Windows version installation to cover the case where users don't uninstall things properly.   What a pain.

Comment (rob-v) [inline OpenRA.Game/Game.cs]: You could just remove ModRegistration from ClearInvalidRegistrations to make it simpler/clearer and keep it as is as it is quite complex (more versions, more system folders,...) for such not gameplay feature.

Comment (rob-v) [inline OpenRA.Game/Game.cs]: But it is now more consistent and ready for future so might be better.

Comment (pchote) [inline OpenRA.Game/Game.cs]: Modifying or removing the system-level registrations require elevated privileges because the files contain executable paths that could otherwise be modified to perform an arbitrary code exploit.

Comment (rob-v) [inline OpenRA.Game/Game.cs]: I meant removing argument ModRegistration registration and Platform.SystemSupportDir from ClearInvalidRegistrations, not removing system registrations. but it is also ok to keep though never used.

**Comment by pchote** [comment]
Added the `--clear-invalid-mod-registrations` utility command for use by the Windows installer.  I think this should finally be ready to go.

Updated test installers:
https://www.dropbox.com/s/0ui7c6274cc16tx/OpenRA-modregistration-20170629.exe?dl=0
https://www.dropbox.com/s/04o5fvqhas1sh88/openra_modregistration.20170629_all.deb?dl=0

File: OpenRA.Game/ExternalMods.cs [hunk 16]
```diff
@@ -38,38 +40,37 @@ public class ExternalMods : IReadOnlyDictionary<string, ExternalMod>
...
+			var sources = new[] { Platform.SystemSupportDir, Platform.SupportDir };
+			foreach (var source in sources.Distinct())
 			{
-				try
-				{
-					var yaml = MiniYaml.FromStream(File.OpenRead(path), path).First().Value;
-					LoadMod(yaml);
-				}
-				catch (Exception e)
+				var supportPath = Path.Combine(source, "ModMetadata");
```
Comment (rob-v) [inline OpenRA.Game/ExternalMods.cs]: nit: called metadataPath elsewhere in this PR.

File: OpenRA.Game/UtilityCommands/ClearInvalidModRegistrationsCommand.cs [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+			if (args[1] == "system" || args[1] == "both")
+				type |= ModRegistration.System;
+
+			if (args[1] == "user" || args[1] == "both")
+				type |= ModRegistration.User;
+
+			var mods = new ExternalMods();
+			ExternalMod activeMod;
+			if (!mods.TryGetValue(ExternalMod.MakeKey(utility.ModData.Manifest), out activeMod))
+				activeMod = null;
```
Comment (rob-v) [inline OpenRA.Game/UtilityCommands/ClearInvalidModRegistrationsCommand.cs]: nit: 'if' not needed, activeMod is null.

**Review by rob-v — APPROVED** [review]
👍 
2 small nits, that aren't worth the troubles with commits and are basically ok. I expect confirmation the test setup issue is unrelated.

**Comment by pchote** [comment]
> I expect confirmation the test setup issue is unrelated.
I can confirm that the crashes happen with a packaged version of bleed, which means that it is either something weird with the system I packaged it on (likely) or some earlier change on bleed (also likely).  Another topic for another issue/PR.

**Comment by RoosterDragon** [comment]
In all the places doing `sources.Distinct()` on paths, do we need to consider using a case-insensitive string comparer on Windows?

**Comment by pchote** [comment]
Shouldnt matter in this case since we have full control over the paths going in to sources

**Comment by pchote** [comment]
Fixed nits and rebased.

**Review by reaperrr — APPROVED** [review]
Couldn't spot any issues (aside from a crash caused by the StyleCop issue, which isn't related to this PR), installation and ModMetadata are added and removed properly by the (un)installer.

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e1df0cb3473e027161c8756e4a0a5cd57d2526e7...7f9f20aa642437a65a605f769062004bb8b3ace0

**Comment by reaperrr** [comment]
Sorry, I only noticed this now, but upon uninstall, it didn't remove the launcher icons from desktop (on Windows). I tested an earlier AppVeyor artifact from a different branch for comparison and there it worked, so it must have regressed here.
